File: 2019/day01/src/main.jl
#!/bin/env julia

function part1(elems)
    [(e ÷ 3 - 2) for e in elems] |> sum
end

function helper(elem)
    elem = (elem ÷ 3) - 2
    ans = elem
    while ans >= 0
        ans = (ans ÷ 3) - 2
        if ans > 0
            elem += ans;
        end # if
    end
    elem
end # function

function part2(elems)
    [helper(e) for e in elems] |> sum
end

function main()
    if isempty(ARGS)
        println(stderr, "Usage:\n\tjulia main.jl <INPUT_FILE>\nOR\n\tmain.jl <INPUT_FILE>")
        exit(1)
    else
        elems = [parse(Int64, e) for e in read(ARGS[1], String) |> split]
    end

    println("Part 1: $(part1(elems))")
    println("Part 1: $(part2(elems))")
end

main()

File: 2019/day01/src/main.rs
use std::{error::Error, fs::File, io::Read, env};

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<_> = env::args().collect();
    if args.len() == 2 {
        let mut file = File::open(&args[1])?;
        let mut input = String::new();
        file.read_to_string(&mut input)?;

        println!("Part 1: {}", part1(&input));
        println!("Part 2: {}", part2(&input));
    } else {
        eprintln!("Usage {} <INPUT_FILE>", args[0]);
    }

    Ok(())
}

fn part1(str: &str) -> i128 {
    str.split_whitespace().map(|e| e.parse::<i128>().unwrap()).map(|e| (e / 3) - 2).sum()
}

fn part2(str: &str) -> i128 {
    str.split_whitespace()
        .map(|e| e.parse::<i128>().unwrap())
        .map(|e| (e / 3) - 2)
        .map(|mut e| {
            let mut ans = e;
            while ans >= 0 {
                ans = (ans / 3) - 2;
                if ans > 0 {
                    e += ans;
                }
            }
            e
        })
        .sum()
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        assert_eq!(2, part1("12"));
        assert_eq!(2, part1("14"));
        assert_eq!(654, part1("1969"));
        assert_eq!(33583, part1("100756"));
    }

    #[test]
    fn test_part2() {
        assert_eq!(2, part2("12"));
        assert_eq!(2, part2("14"));
        assert_eq!(966, part2("1969"));
        assert_eq!(50346, part2("100756"));
    }
}
